### PR TITLE
RUMM-308 Make Logger methods Thread safe

### PR DESCRIPTION
**Pull request by mariusc83**
### What does this PR do?

In this PR we are fixing the Logger **remove/add** methods for tags or attributes by making them Thread safe. To achieve this we are switching from regular HashMap and HashSet implementations to a ConcurrentHashMap and a CopyOnWriteArraySet implementations.

### Additional Notes

Please note that the problem was also present in Android as I was able to reproduce the **ConcurrentModificationException** in my tests.

### Review checklist (to be filled by reviewers)

- [x] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [x] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [x] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

File: dd-sdk-android/src/main/kotlin/com/datadog/android/log/Logger.kt [hunk 1]
```diff
@@ -466,5 +468,21 @@ internal constructor(private val handler: LogHandler) {
...
+            attributes[key] = value
+        }
+    }
+
+    private fun safelyRemoveTagsWithKey(keyFilter: (String) -> Boolean) {
+        // we first gather all the objects we want to remove based on a copy
+        val toRemove = tags
+            .toArray()
+            .filter {
+                it is String && keyFilter(it)
```
Comment (xgouchet) [inline dd-sdk-android/src/main/kotlin/com/datadog/android/log/Logger.kt]: in theory it is always string ! 
Why can't you just use `tags.filter(keyFilter)` ?

File: dd-sdk-android/src/test/kotlin/com/datadog/android/log/LoggerTest.kt [hunk 2]
```diff
@@ -366,7 +370,7 @@ internal class LoggerTest {
                 Log.INFO,
                 fakeMessage,
                 null,
-                mapOf(key to value),
+                emptyMap(),
```
Comment (xgouchet) [inline dd-sdk-android/src/test/kotlin/com/datadog/android/log/LoggerTest.kt]: You're breaking the way logs work here. When they look the logs attributes online, instead of seeing `my_attribute: null`, they won't see `my_attribute` at all :/

File: dd-sdk-android/src/test/kotlin/com/datadog/android/log/LoggerTest.kt [hunk 3]
```diff
@@ -833,4 +837,114 @@ internal class LoggerTest {
     }
 
     // endregion
+
+    // region Multi Thread Access
+
+    @Test
+    fun `adding tags or attributes is thread safe`(forge: Forge) {
```
Comment (xgouchet) [inline dd-sdk-android/src/test/kotlin/com/datadog/android/log/LoggerTest.kt]: This test method is huge. Can you break it in two, one to test tags, and one to test attributes. 
Maybe try to make the code a bit more readable, it's hard to follow your logic here.

File: dd-sdk-android/src/main/kotlin/com/datadog/android/log/Logger.kt (hunk 1 above)
Comment (mariusc83) [reply dd-sdk-android/src/main/kotlin/com/datadog/android/log/Logger.kt]: if you do that will not be thread safe anymore. You need to use `toArray` function to make a copy because otherwise another thread might modify the `collection` while you are filtering (ConcurrentModificationException). Problem is that `toArray` returns an Object[] for some reason and I need it to cast it to a String there. The method I am using after `removeAll` it is already Thread safe on at the collection level (it uses a synchronized block).

File: dd-sdk-android/src/test/kotlin/com/datadog/android/log/LoggerTest.kt (hunk 3 above)
Comment (mariusc83) [reply dd-sdk-android/src/test/kotlin/com/datadog/android/log/LoggerTest.kt]: Agree...good idea to break it in 2 for tags and attributes.

File: dd-sdk-android/src/test/kotlin/com/datadog/android/log/LoggerTest.kt (hunk 2 above)
Comment (mariusc83) [reply dd-sdk-android/src/test/kotlin/com/datadog/android/log/LoggerTest.kt]: This collection does not accept "null" as values and it will throw an NPE, what if instead of NULL we add a "null" as String ?

Comment (xgouchet) [reply dd-sdk-android/src/test/kotlin/com/datadog/android/log/LoggerTest.kt]: That won't work, since `"null"` != `null`
What you can do is create a `val NULL = Object()`, and pass it instead, and when serializing convert that to JsonNull.

File: dd-sdk-android/src/main/kotlin/com/datadog/android/log/Logger.kt (hunk 1 above)
Comment (xgouchet) [reply dd-sdk-android/src/main/kotlin/com/datadog/android/log/Logger.kt]: Okay, makes sense, but then you can use Kotlin's `toTypedArray()` instead, it will turn it into a `String[]`
